### PR TITLE
Handle network errors and retry upload/download data with backoff to/from object storage

### LINKED ISSUE #318: Retry on network failures (e.g uploads)
Not critical since compactor just restarted and continued just fine, but can be annoying.

```
level=error name=thanos-compactor ts=2018-04-28T10:32:12.73383864Z caller=main.go:147 msg="running command failed" err="first pass of downsampling failed: retrieve bucket block metas: get meta for block 01C6XZ1256S7VFNQP9D36XJ4F4: Get https://storage.googleapis.com/thanos-alpha/01C6XZ1256S7VFNQP9D36XJ4F4/meta.json: dial tcp [xxx]:443: connect: network is unreachable"
```

### PR DESCRIPTION
**Pull request by xjewer**
Add backoff reply for a single object storage request, except Range and Iter.
Error handler splits errors on net/http and others, and replies the request to the object storage for the former.

Fixes #318

cc @bwplotka

### CODE REVIEW

File: docs/components/compact.md [hunk 1]
```diff
@@ -60,7 +60,7 @@ Flags:
       --retention.resolution-1h=0d  
                            How long to retain samples of resolution 2 (1 hour)
                            in bucket. 0d - disables this retention
-  -w, --wait               Do not exit after all compactions have been processed
-                           and wait for new work.
+  -r, --repeat=5m          Do not exit after all compactions have been processed
```
Comment (bwplotka) [inline docs/components/compact.md]: Can we move the name in different PR and discuss about it? Sure we are in minor release but we cannot break compatiblity that easily in every PR. Someday there will be another contributor who will change it in random PR to something else. This one makes sense, just we need to be nice to users and not break it. Plus changelog change is must have here.

File: cmd/thanos/compact.go [hunk 2]
```diff
@@ -178,12 +179,12 @@ func runCompact(
...
 
-		if !wait {
+		if repeat == 0 {
 			return f()
 		}
 
-		// --wait=true is specified.
-		return runutil.Repeat(5*time.Minute, ctx.Done(), func() error {
+		// --repeat is specified.
+		return runutil.Repeat(backoff.NewConstantBackOff(repeat), ctx.Done(), func() error {
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: I think you confused `Repeat` with `Retry` 

We don't want to have backoff here. This is just periodic compaction+downsampling+retention run that is periodically ran... This is executed for both retry one error, sure, but also for normat thing.

File: pkg/store/bucket.go [hunk 3]
```diff
@@ -1627,3 +1627,25 @@ func (s queryStats) merge(o *queryStats) *queryStats {
 
 	return &s
 }
+
+// GetMetas gets all bucket metas
+func GetMetas(ctx context.Context, logger log.Logger, bkt objstore.Bucket) ([]*block.Meta, error) {
```
Comment (bwplotka) [inline pkg/store/bucket.go]: Why do you think this name is better? I feel personally that `DownloadMetas` makes more sense, as you show that it is complex thing, not getting something from memory. 

Also getters for Golang never have `Get` prefix and it looks like a getter with this prefix (:

Comment (xjewer) [reply pkg/store/bucket.go]: Download means we put this data on our local disk, so here that's confusing, imo, Get is more suitable, the same as the method `Get` in the `BucketReader` interface

File: cmd/thanos/compact.go (hunk 2 above)
Comment (xjewer) [reply cmd/thanos/compact.go]: Actually not, my thought was use the same interface to pass time intervals to Repeat or Retry methods. 

So further, we can pass custom backoff, not only constant, but the customizable.
But the term of backoff seems not to be clear, indeed.

File: docs/components/compact.md (hunk 1 above)
Comment (xjewer) [reply docs/components/compact.md]: >This one makes sense, just we need to be nice to users and not break it.
So you suggest to leave `--wait` with a default 5m interval here?

I understand you concern, but splitting this changes it 2 requests looks not to be convenient. PR has exactly the purpose to discuss changes and find tradeoff 😀

File: cmd/thanos/compact.go (hunk 2 above)
Comment (bwplotka) [reply cmd/thanos/compact.go]: Yup, got confused. ^^ Constant backoff like a  "hot ice"  or "moving constant" opposite words.

Also not sure any backoff is nice for repeat, ticker based function. Maybe we need to fix just naming here

File: pkg/objstore/metric_bucket.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// BucketWithMetrics takes a bucket and registers metrics with the given registry for
```
Comment (bwplotka) [inline pkg/objstore/metric_bucket.go]: What is this, how this is related to the adding retries and backoff?

File: docs/components/compact.md (hunk 1 above)
Comment (bwplotka) [reply docs/components/compact.md]: yea.. maybe you are right just I was overwhelmed how many things you are changing here just to fix retries. So .. let's discuss it here (: 

When you think about `repeat` flag. I think we don't need that one. Why you even care how often it is running? IMO there is no reason to control this. Compactor should basically run all the time, maybe even as often as needed. So I would not change this. What do you think? (:

File: pkg/store/bucket.go (hunk 3 above)
Comment (bwplotka) [reply pkg/store/bucket.go]: hm, can agree

File: pkg/objstore/metric_bucket.go (hunk 4 above)
Comment (bwplotka) [reply pkg/objstore/metric_bucket.go]: hm?

Comment (xjewer) [reply pkg/objstore/metric_bucket.go]: I just moved it from objstore.go https://github.com/improbable-eng/thanos/blob/master/pkg/objstore/objstore.go#L172, as it's a separate entity

File: cmd/thanos/compact.go (hunk 2 above)
Comment (xjewer) [reply cmd/thanos/compact.go]: fixing naming of backoff? that's a external package.
I can't leave it as it was to avoid confusion with backoff vs repeat (not retry)

Comment (bwplotka) [reply cmd/thanos/compact.go]: yup

File: docs/components/compact.md (hunk 1 above)
Comment (xjewer) [reply docs/components/compact.md]: anyway, even if you don't need customise compactor (though for the small setup no need to scan blocks' meta every 5 minutes, you can do it once per hours), the `wait` flag seems not to be appropriate in conjunction with time intervals passed to it

**Comment by bwplotka** [comment]
OK I dived deeper into minio client and there are definitely some retries there:
https://sourcegraph.com/github.com/minio/minio-go@master/-/blob/api.go#L524:17

I guess @xjewer you seen errors on this s3 part? That requires retries? If that is true we have to ask them how to deal with those, as it seems like their retry part is not perfect. 

Went through the idea with @fabiansiddiqia and we both think retry has to be part of client implementation for each one separatedly. If you use S3, feel free to focus on making sure S3 client retries properly.

Sorry for having this delay and thanks for your code here, but we have to chase minio-go for this and check if we can tune the client retry args (if any).

**Comment by xjewer** [comment]
>OK I dived deeper into minio client and there are definitely some retries there:
https://sourcegraph.com/github.com/minio/minio-go@master/-/blob/api.go#L524:17

>I guess @xjewer you seen errors on this s3 part? That requires retries? If that is true we have to ask them how to deal with those, as it seems like their retry part is not perfect.

The PR itself not only about S3, no other implementations doesn't care about retries.

>I feel like we retry on every error which might be not ok.. cannot see the code for checking if it is only network error

https://github.com/improbable-eng/thanos/pull/537/files#diff-adfb3baee650b10fb14c291c17b078fcR351

>Because simply is impossible for consumer of intereface to tell on which error should be retriable without knowing implementation and if the retry was NOT already there for some implementation.

I can agree, that this handler could be moved into the implementations and let's them return proper type of error: permanent or retriable. 

> This is abstraction leak to me. You can mitigate this by adding special method to interface like IsErrRetriable but it might unnecessary increase complexity of Bucket interface and make interface shallow.

We already have [metricBucket](https://github.com/improbable-eng/thanos/blob/master/pkg/objstore/objstore.go#L208) proxy and recently [removed](https://github.com/improbable-eng/thanos/pull/599) internal metrics from implementations, what's the difference with another proxy to implement backoff in one place?

if there's a wish to move it behind implementation, there're 2 ways to go:
1. just agree with storage maintainers https://github.com/improbable-eng/thanos/blob/master/docs/storage.md not to implement logic, but just return the proper error
2. raise a issue to solve retry logic for each object storage client (minio client definitely doesn't work well with it, so should be fixed)

@bwplotka @vglafirov @sudhi-vm @jojohappy (with https://github.com/improbable-eng/thanos/pull/467)

**Comment by bwplotka** [comment]
> We already have metricBucket proxy and recently removed internal metrics from implementations, what's the difference with another proxy to implement backoff in one place?

The difference is that:
* metric bucket does not care what is the implementation. It reports metrics per our interface usage. So we can make this simplification.
* Our Implementations already have some kind of retries e.g minio + HTTP retries. If they work.. that's another story (:
* Even if we add IsRetryError interface, what if library itself does not give us this info, because error is masked? What if we will use https://github.com/google/go-cloud/blob/master/blob/blob.go someday? 

> raise a issue to solve retry logic for each object storage client (minio client definitely doesn't work well with it, so should be fixed)
IMO that's the solution. Having that will allow you to not add another interface for Iter (ListNames)

**Comment by bwplotka** [comment]
Hey, you get any luck in asking minio about retries for different API methods? I think that's the right direction to start with

**Comment by bwplotka** [comment]
Looks like the agreed idea is to fix retries per component. Minio and GCS has retries built in. 

Thanks for this work anyway!
